pilot/pkg/networking/core/v1alpha3/cluster_test.go

```diff
@@ -3277,3 +3278,479 @@ func TestEnvoyFilterPatching(t *testing.T) {
+										Values: []*structpb.Value{
+											{
+												Kind: &structpb.Value_StructValue{
+													StructValue: &structpb.Struct{
+														Fields: map[string]*structpb.Value{
+															"host": {
+																Kind: &structpb.Value_StringValue{
+																	StringValue: "a.default",
+																},
+															},
+															"name": {
+																Kind: &structpb.Value_StringValue{
+																	StringValue: "a",
+																},
+															},
+															"namespace": {
+																Kind: &structpb.Value_StringValue{
+																	StringValue: "default",
+																},
+															},
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:      "outbound service metadata",
+			direction: model.TrafficDirectionOutbound,
+			cluster:   &cluster.Cluster{},
+			service: &model.Service{
+				Attributes: model.ServiceAttributes{
+					Name:      "a",
+					Namespace: "default",
+				},
+				Hostname: "a.default",
+			},
+			want: &core.Metadata{
+				FilterMetadata: map[string]*structpb.Struct{
+					util.IstioMetadataKey: {
+						Fields: map[string]*structpb.Value{
+							"services": {
+								Kind: &structpb.Value_ListValue{
+									ListValue: &structpb.ListValue{
+										Values: []*structpb.Value{
+											{
+												Kind: &structpb.Value_StructValue{
+													StructValue: &structpb.Struct{
+														Fields: map[string]*structpb.Value{
+															"host": {
+																Kind: &structpb.Value_StringValue{
+																	StringValue: "a.default",
+																},
+															},
+															"name": {
+																Kind: &structpb.Value_StringValue{
+																	StringValue: "a",
+																},
+															},
+															"namespace": {
+																Kind: &structpb.Value_StringValue{
+																	StringValue: "default",
+																},
+															},
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:      "inbound duplicated metadata",
+			direction: model.TrafficDirectionInbound,
+			cluster:   &cluster.Cluster{},
+			svcInsts: []*model.ServiceInstance{
+				{
+					Service: &model.Service{
+						Attributes: model.ServiceAttributes{
+							Name:      "a",
+							Namespace: "default",
+						},
+						Hostname: "a.default",
+					},
+					ServicePort: &model.Port{
+						Port: 80,
+					},
+				},
+				{
+					Service: &model.Service{
+						Attributes: model.ServiceAttributes{
+							Name:      "a",
+							Namespace: "default",
+						},
+						Hostname: "a.default",
+					},
+					ServicePort: &model.Port{
+						Port: 80,
+					},
+				},
+			},
+			want: &core.Metadata{
+				FilterMetadata: map[string]*structpb.Struct{
```

Comment:
readability nit: This could be more readable if 
```go
want: &core.FilterMetadata: ParseYamlToStruct('''
services: [ {host: a.default, name: a, namespace: default} ]
'''
 }

Reply:
I searched a bit and can not find a convenient util like this. Parsing a json object would work, but the readability is equally bad.

Reply:
Fine not to do it.
I totally disagree that readability it equally bad. A terse json or yaml snippet that lets you concentrate on data is better than something that nested 8 levels deep.

Reply:
you can use `jsonpb.UnmarshalJson(json, &struct)` I think. or similar